plugins/modules/katello_manifest.py

```diff
@@ -76,116 +61,71 @@
-            changed = True
-    return changed
+from ansible.module_utils.foreman_helper import KatelloEntityApypieAnsibleModule
 
 
 def main():
-    module = AnsibleModule(
+    module = KatelloEntityApypieAnsibleModule(
         argument_spec=dict(
-            server_url=dict(required=True),
-            username=dict(required=True, no_log=True),
-            password=dict(required=True, no_log=True),
-            verify_ssl=dict(type='bool', default=True),
-            organization=dict(required=True),
             manifest_path=dict(),
             state=dict(default='present', choices=['absent', 'present', 'refreshed']),
-            redhat_repository_url=dict(),
+            repository_url=dict(aliases=['redhat_repository_url']),
         ),
         required_if=[
             ['state', 'present', ['manifest_path']],
         ],
-        supports_check_mode=True,
     )
 
-    if has_import_error:
-        module.fail_json(msg=import_error_msg)
+    entity_dict = module.clean_params()
 
-    set_task_timeout(300000)  # 5 minutes
+    module.connect()
 
-    server_url = module.params['server_url']
-    username = module.params['username']
-    password = module.params['password']
-    verify_ssl = module.params['verify_ssl']
-    organization = module.params['organization']
-    manifest_path = module.params['manifest_path']
-    redhat_repository_url = module.params['redhat_repository_url']
-    state = module.params['state']
+    organization = module.find_resource_by_name('organizations', name=entity_dict['organization'], thin=False)
+    scope = {'organization_id': organization['id']}
 
-    create_server(server_url, (username, password), verify_ssl)
-    ping_server(module)
+    existing_manifest = organization.get('owner_details', {}).get('upstreamConsumer')
+
+    changed = False
+    if module.state == 'present':
+        if 'repository_url' in entity_dict:
+            payload = {'redhat_repository_url': entity_dict['repository_url']}
+            org_spec = dict(redhat_repository_url=dict())
+            changed_url, organization = module.ensure_entity('organizations', payload, organization, state='present', entity_spec=org_spec)
+        else:
+            changed_url = False
+
+        try:
+            with open(entity_dict['manifest_path'], 'rb') as manifest_file:
+                files = {'content': (entity_dict['manifest_path'], manifest_file, 'application/zip')}
+                headers = {'content_type': 'multipart/form-data', 'multipart': 'true'}
+                params = scope.copy()
+                if 'repository_url' in entity_dict:
+                    params['repository_url'] = entity_dict['repository_url']
+                changed, result = module.resource_action('subscriptions', 'upload', params, options={'skip_validation': True}, files=files)
+                task = module.wait_for_task(result, 300)  # 5 minutes
+                for error in task['humanized']['errors']:
```

Comment:
Should we manage errors in exceptions?
We can postpone this.

Reply:
Not sure I follow what you mean? `wait_for_tasks` does not raise anything at the moment.

Reply:
In my suggestion I was thinking about `raise TimeoutError` instead of `self.fail_json`

Reply:
The question was whether it should raise exceptions, which would make handling them a bit more comfortable.

Reply:
I never understood, why ansible does not like exceptions for error-reporting.

Reply:
Ah, OK, now I get it. Yeah we can make it raise various Exceptions and handle these here.

Not sure there is an added benefit though (besides being more Pythonic than Ansiblig)